query_optimizer/logical/Logical.hpp

```diff
@@ -64,12 +70,62 @@ class Logical : public OptimizerTree<Logical> {
    */
   virtual std::vector<expressions::AttributeReferencePtr> getReferencedAttributes() const = 0;
 
+  /**
+   * @brief Make a copy of this logical node with input expressions replaced by
+   *        input_expression.
+   *
+   * @return A copy of this logical node with the new input expressions.
+   */
+  virtual LogicalPtr copyWithNewInputExpressions(
+      const std::vector<expressions::ExpressionPtr> &input_expressions) const {
+    // Default implementation: return a shared reference to "this" when there is
+    // no change.
+    DCHECK(input_expressions.empty());
+    return shared_from_this();
+  }
+
+  /**
+   * @return The input expressions in this Logical.
+   */
+  const std::vector<expressions::ExpressionPtr>& input_expressions() const {
+    return input_expressions_;
+  }
+
+  /**
+   * @return The input attributes, which are the union set of the output attributes of
+   *         all children.
+   */
+  std::vector<expressions::AttributeReferencePtr> getInputAttributes() const {
+    std::vector<expressions::AttributeReferencePtr> input_attributes;
+    for (const LogicalPtr &child : children()) {
+      const std::vector<expressions::AttributeReferencePtr> child_output_attributes =
```

Comment:
I think typically there are three scenarios.

Senario 1:
```
A a;
A b = a;
```
In this case, `A b = a` is equivalent to `A b(a)` which will invoke `b`'s **copy constructor** with `a` as parameter.

Scenario 2:
```
A a;
A b = std::move(a);
```
In this case, `A b = std::move(a)` is equivalent to `A b(std::move(a))` which will invoke `b`'s **move constructor** with `std::move(a)` as parameter.

Scenario 3:
```
A f() {
  A a;
  return a;
}

void main() {
  A b = f();
}
```
In this scenario, typically if the `A`'s constructor has no side effect, then there will be **neither copy construction nor move construction**, known as the Named Return Value Optimization (NRVO).